### PR TITLE
GO-3913 separate getting identities into multiple calls to identity

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
Description: https://linear.app/anytype/issue/GO-3913/separate-getting-identities-into-multiple-calls-to-identity-repo

### CODE REVIEW

File: core/identity/identity.go [hunk 1]
```diff
@@ -280,6 +278,32 @@ func (s *service) getIdentitiesDataFromRepo(ctx context.Context, identities []st
...
+			repoIdentities      []*identityrepoproto.DataWithIdentity
+			err                 error
+			requestedIdentities []string
+		)
+		if j+identityBatch < len(identities) {
+			requestedIdentities = identities[j : j+identityBatch]
+		} else {
+			requestedIdentities = identities[j:]
+		}
+		repoIdentities, err = s.identityRepoClient.IdentityRepoGet(ctx, requestedIdentities, []string{identityRepoDataKind})
```
Comment (deff7) [inline core/identity/identity.go:295]: maybe send requests in parallel?

File: core/identity/identity.go [hunk 2]
```diff
@@ -280,6 +278,32 @@ func (s *service) getIdentitiesDataFromRepo(ctx context.Context, identities []st
...
+		if err != nil {
+			log.Error("error getting identity repo data", zap.Error(err))
+			failedIdentities = append(failedIdentities, requestedIdentities...)
+		} else {
+			resultIdentities = append(resultIdentities, repoIdentities...)
+		}
+		j += identityBatch
```
Comment (deff7) [inline core/identity/identity.go:302]: maybe move it to the loop expression?

```
for j := 0; j < len(identities); j+= identityBatch {
```

File: core/identity/identity.go [hunk 3]
```diff
@@ -280,6 +278,32 @@ func (s *service) getIdentitiesDataFromRepo(ctx context.Context, identities []st
...
+			requestedIdentities []string
+		)
+		if j+identityBatch < len(identities) {
+			requestedIdentities = identities[j : j+identityBatch]
+		} else {
+			requestedIdentities = identities[j:]
+		}
+		repoIdentities, err = s.identityRepoClient.IdentityRepoGet(ctx, requestedIdentities, []string{identityRepoDataKind})
+		if err != nil {
+			log.Error("error getting identity repo data", zap.Error(err))
```
Comment (deff7) [inline core/identity/identity.go:297]: I'm concerned about the number of log messages. Is it possible that we will log unnecessary errors?

File: core/identity/identity.go [hunk 4]
```diff
@@ -280,6 +278,32 @@ func (s *service) getIdentitiesDataFromRepo(ctx context.Context, identities []st
 	return res, nil
 }
 
+func (s *service) batchIdentities(ctx context.Context, identities []string) ([]*identityrepoproto.DataWithIdentity, []string) {
+	resultIdentities := make([]*identityrepoproto.DataWithIdentity, 0, len(identities))
```
Comment (mcrakhman) [inline core/identity/identity.go:282]: Just summing all the comments, there should be one abstract function, which splits the []T to several batches of size <= N (take N as 200 pls), this function should be in package `slice` or whatever and tested

and then call in parallel some named func for every batch which returns results using waitgroup, gather everything and return.

This func should either get stuff from remote or if nothing in remote get stuff from local.

I would do it like this

File: core/identity/identity.go (hunk 3 above)
Comment (AnastasiaShemyakinskaya) [reply core/identity/identity.go:297]: May be we can collect such errors and then log it in one log? 🧐

**Review by mcrakhman — APPROVED** [review]

**Review by mcrakhman — APPROVED** [review]

**Review by mcrakhman — APPROVED** [review]

**Pull request merged**